### PR TITLE
Support reading CHDv5

### LINKED ISSUE #18: Update chd/chdr/chdv5 code
@rtissera has made an updated lib from the old chd code that supports v5. We should switch to it - https://github.com/rtissera/libchdr

Related: #676 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31269190-update-chd-chdr-chdv5-code?utm_campaign=plugin&utm_content=tracker%2F500311&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F500311&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by baka0815**
This imports https://github.com/rtissera/libchdr/commit/6726f52a1d3ba04cebefade2bfcbb9dfc85eef00 and also the open pull request (https://github.com/rtissera/libchdr/pull/8 and https://github.com/rtissera/libchdr/pull/12).

I tested the following settings (on Linux and Windows) using a verified TOSEC dump of Crazy Taxy  (PAL):
* ``cdfl (CD FLAC)``
* ``cdlz (CD LZMA)``
* ``cdzl (CD Deflate)``
* ``zlib (Deflate)``
* ``none``

and the default
* ``cdlz (CD LZMA), cdzl (CD Deflate), cdfl (CD FLAC)``

All were working flawlessly.

What's not (yet) working are the following codecs, however they can be added later (like we did with ``zlib (Deflate)``)
* ``flac (FLAC)``
* ``huff (Huffman)``
* ``lzma (LZMA)``

This fixes #18

### CODE REVIEW

**Comment by skmp** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 84b17fea8a7d830bba7bdcb3bf107917e000cb72 into a8711e2cf253085edab3e3b95e452cb633ad9211 - [view on LGTM.com](https://lgtm.com/projects/g/reicast/reicast-emulator/rev/pr-e17625ed3ad03d2043ab41119c1fc1f73b0841e3)

**new alerts:**

* 1 for FIXME comment

**fixed alerts:**

* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by skmp** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 67ccc1aa3afebe984d96d59cf14842bb7c0662e6 into a8711e2cf253085edab3e3b95e452cb633ad9211 - [view on LGTM.com](https://lgtm.com/projects/g/reicast/reicast-emulator/rev/pr-2155b5b415131d18670244ee94b3d1b39867cdf6)

**new alerts:**

* 1 for FIXME comment

**fixed alerts:**

* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by baka0815** [comment]
@skmp is there anything blocking this?

@AbandonedCart do you have time to review this for Android?

**Comment by baka0815** [comment]
Ok, that **should** work, at least it does for me locally.
However it may be that FLAC breaks again - maybe the libs aren't installed on travis and we need to do that?

**Comment by baka0815** [comment]
Yep, the wercker needs libflac-dev.

**Comment by skmp** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 20ff338283c4bab18189f56c98f113ae0425db81 into 00ecf2f1cecc022fae693032f8dd382ea719d8e3 - [view on LGTM.com](https://lgtm.com/projects/g/reicast/reicast-emulator/rev/pr-3c1e261c3ac01751bef186d058618154e9932062)

**new alerts:**

* 1 for FIXME comment

**fixed alerts:**

* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by baka0815** [comment]
Poking @luserx0 by @skmp's request.

**Comment by baka0815** [comment]
Ok, I checked on Android and ``cdzl (CD Deflate)``, ``zlib (Deflate)`` and ``none`` are working with Crazy Taxi.
However ``cdfl (CD FLAC)`` and ``cdlz (CD LZMA)`` are not, so there might be problems with LZMA and FLAC on Android. Could we fix them later?

What I currently don't understand is that some games are not working (like F355, see https://github.com/libretro/reicast-emulator/issues/54#issuecomment-418874079).

**Comment by baka0815** [comment]
Ok, there seems to be more to it.
I tested the following games with chdman 0.200 and the standard compression (cdlz (CD LZMA), cdzl (CD Deflate), cdfl (CD FLAC)). Some are working, many aren't.

__Working games__
* Crazy Taxi v1.000 (2000)(Sega)(PAL)[!]
* Dino Crisis v1.000 (2000)(Virgin)(PAL)[!]

__Nonworking games__
* 102 Dalmatians - Puppies to the Rescue v1.000 (2000)(Eidos)(PAL)(GB)[!]
* 102 Dalmatians - Puppies to the Rescue v1.000 (2001)(Eidos)(PAL)(M4)[!]
* Alone in the Dark - The New Nightmare v1.000 (2001)(Infogrames)(PAL)(Disc 1 of 2)[!]
* Blue Stinger v1.000 (1999)(Activision)(PAL)[!]
* Bust-A-Move 4 v1.001 (2000)(Acclaim)(PAL)[!]
* Capcom vs. SNK v1.000 (2000)(Virgin)(PAL)[!]
* Confidential Mission v1.002 (2001)(Sega)(PAL)(M5)[!]
* Daytona USA 2001 v1.002 (2001)(Sega)(PAL)(M4)[!]
* Dead or Alive 2 v1.000 (2000)(Acclaim)(PAL)[!]
* Ecco the Dolphin - Defender of the Future v1.003 (2000)(Sega)(PAL)(M4)[!]
* F355 Challenge - Passione Rossa v1.001 (2000)(Acclaim)(PAL)[!]
* Grand Theft Auto 2 v1.002 (2000)(Rockstar)(US)[!]
* Jet Grind Radio v1.005 (2000)(Sega)(US)[!]
* Ooga Booga v1.001 (2001)(Sega)(US)[!]
* Project Justice - Rival Schools 2 v1.001 (2000)(Virgin)(PAL)[!]
* Rayman 2 - The Great Escape v1.003 (2000)(Ubi Soft)(US)(M5)[!]
* Re-Volt v1.001 (1999)(Acclaim)(PAL)(M5)[!]

But this has nothing to do with the compression (LZMA, FLAC, etc.) as "none" and "zlib (Deflate") are also not working for the nonworking cases.

**Comment by baka0815** [comment]
Here is the diff between Raycast 2 as GDI and CHD (I removed the "reicast git/n" entries).
There is a new ta context created using the GDI which is missing in the CHD version, however the loading looks fine.
I have no idea currently what could be the cause here.
````diff
12c12
< vmem reserve: base: CB360000, aram: eb360000, vram: CF360000, ram: D7360000
---
> vmem reserve: base: B4820000, aram: d4820000, vram: B8820000, ram: C0820000
14c14
< Mem alloc successful!Using 'Rayman 2 - The Great Escape v1.003 (2000)(Ubi Soft)(US)(M5)[!].gdi' as cd image
---
> Mem alloc successful!Using 'v5.chd' as cd image
34,39c34,38
< 
< GDI : 3 tracks
< file[1] "track01.bin": FAD:0, CTRL:4, SSIZE:2352, OFFSET:0
< file[2] "track02.raw": FAD:721, CTRL:0, SSIZE:2352, OFFSET:0
< file[3] "track03.bin": FAD:45000, CTRL:4, SSIZE:2352, OFFSET:0
< gdrom: Opened image "Rayman 2 - The Great Escape v1.003 (2000)(Ubi Soft)(US)(M5)[!].gdi"
---
> chd: parsing file v5.chd
> TRACK:1 TYPE:MODE1_RAW SUBTYPE:NONE FRAMES:721 PAD:150 PREGAP:0 PGTYPE:MODE1 PGSUB:NONE POSTGAP:0
> TRACK:2 TYPE:AUDIO SUBTYPE:NONE FRAMES:44279 PAD:43783 PREGAP:0 PGTYPE:MODE1 PGSUB:NONE POSTGAP:0
> TRACK:3 TYPE:MODE1_RAW SUBTYPE:NONE FRAMES:504150 PAD:0 PREGAP:0 PGTYPE:MODE1 PGSUB:NONE POSTGAP:0
> gdrom: Opened image "v5.chd"
47c46
< 	 CodeCache addr: 0x55606f815000 | from: 0x55606f815000 | addr here: 0x55606f8134a0
---
> 	 CodeCache addr: 0x563cd2ab7000 | from: 0x563cd2ab7000 | addr here: 0x563cd2ab54a0
169d167
< new tactx
183c181
< Sh4: i-cache invalidation AC00043A
---
> Sh4: i-cache invalidation 8C00DD78
185c183,187
< recSh4:Dynarec Cache clear at AC00043A
---
> recSh4:Dynarec Cache clear at 8C00DD78
> DIV32S matched 1% @ 0x8C00CF78
> DIV32S matched 100% @ 0x8C00CF7E
> div32s 1/0/3
> Sh4: i-cache invalidation 8C09E6A6
187,188c189,190
< recSh4:Dynarec Cache clear at AC0082FE
< Sh4: i-cache invalidation AC0082FE
---
> recSh4:Dynarec Cache clear at 8C09E6A6
> Sh4: i-cache invalidation AC09D990
190c192
< recSh4:Dynarec Cache clear at AC0082FE
---
> recSh4:Dynarec Cache clear at AC09D990
192,196c194,195
< DIV32S matched 1% @ 0x8C0098D8
< DIV32S matched 100% @ 0x8C0098DE
< div32s 1/0/3
< DIV32S matched 1% @ 0x8C009A36
< DIV32S matched 100% @ 0x8C009A3E
---
> DIV32S matched 1% @ 0x8C091086
> DIV32S matched 100% @ 0x8C09108C
198,215c197,198
< DIV32S matched 1% @ 0x8C009AC0
< DIV32S matched 1% @ 0x8C009ACA
< Sh4: i-cache invalidation AC00E0DA
< Freeing fpcb
< recSh4:Dynarec Cache clear at AC00E0DA
< Freeing fpcb
< recSh4:Dynarec Cache clear at AC00FFFE
< Sh4: i-cache invalidation AC00FFFE
< Freeing fpcb
< recSh4:Dynarec Cache clear at AC00FFFE
< Sh4: i-cache invalidation 8C0D7BD2
< Freeing fpcb
< recSh4:Dynarec Cache clear at 8C0D7BD2
< Sh4: i-cache invalidation AC0CF84E
< Freeing fpcb
< recSh4:Dynarec Cache clear at AC0CF84E
< DIV32S matched 1% @ 0x8C06794A
< DIV32S matched 100% @ 0x8C067950
---
> DIV32S matched 1% @ 0x8C0911E4
> DIV32S matched 100% @ 0x8C0911EC
217,218c200,201
< SPI : unknown ? [0x70]
< SPI : unknown ? [0x71]
---
> DIV32S matched 1% @ 0x8C09126E
> DIV32S matched 1% @ 0x8C091278
222,224c205,206
< FOG TABLE Curve match: maxdev: 100.00 cents
< DIV32S matched 1% @ 0x8C067AA8
< DIV32S matched 100% @ 0x8C067AB0
---
> DIV32S matched 1% @ 0x8C0294A6
> DIV32S matched 100% @ 0x8C0294AC
226c208,211
< DIV32S matched 1% @ 0x8C067B32
---
> SPI : unknown ? [0x70]
> SPI : unknown ? [0x71]
> SPI : unknown ? [0x70]
> SPI : unknown ? [0x71]
````

**Comment by AbandonedCart** [comment]
https://github.com/Chainfire/android-ndk-compression-tools may help with debugging lzma for Android. It's under the xz folder. I can't say how reliable that code is, since the dev had a habit of doing the minimum requirement to profit before selling off the rights.

**Comment by baka0815** [comment]
https://pastebin.com/nViHuLqa

Both have no pre- or postgap
No sub-types and only MODE1_RAW  tracks
The GDI versions of both games are running fine
I double-checked the loading of the metadata, the info printed by chdman is identical to what we have.

Looks like I need help here.

File: core/deps/chdr/coretypes.h [hunk 1]
```diff
@@ -1,18 +1,34 @@
+#ifndef __CORETYPES_H__
```
Comment (skmp) [inline core/deps/chdr/coretypes.h]: `#pragma once` is much preferred over `#ifndef guards` and should be used for new code

**Comment by skmp** [comment]
Alright, did a quick review of this. I'm 99% sure the reason some discs aren't working is the calculation of the TOC / FADs from the chd toc.

A few comments on the PR overall
- It is wayy, wayyy too big. I can't load the diff in the browser. It should be trimmed down to include only the specific files needed from each library, and not things like documentation or examples.
- There are a LOT of whitespace changes. As a rule of thumb, whitespace changes should not be included in diffs that change functionality, as this makes reviewing the diff much harder. White space and formatting changes should be done in separate PRs

Can you iterate on these? One the diff is more manageable, I'll be able to provide more technical feedback on what the issues are %)

Thanks

**Comment by baka0815** [comment]
Ok, I'll close this for now and restart this once the cmake conversion is over.
I'm not entirely happy with how the deps are handled in the Makefile (all *.c* files are included - that's the reason i deleted some).

However I included the whole lib to make updates in the future more easily. So there will be many changes just because of that. That's one reason I did the whole pull in a separatr commit.

I will still try to look after the current issue (maybe toc).

**Comment by baka0815** [comment]
Ok, I found something, but that irritates me.

I took a look at what ReadSector() returns:
https://github.com/reicast/reicast-emulator/blob/eacbae5b30bcf64f75733a4fe761df9aeb64ce71/core/imgread/common.h#L157

And reading the CHD (for F355 at least) was reading different data than the GDI (Crazy Taxi returns the same data).

it looks like hunk_read_into_memory() gets the wrong block offset. The offset is about 9792 off (or hunkbytes / 2).
https://github.com/reicast/reicast-emulator/pull/1363/files#diff-61b5b7740b28fa01a2a98502aef0d2d1R2091

But that's only valid for the uncompressed case and not for Crazy Taxi.
However the loading of the map and the header seems to be fine as I double checked the code with MAME and took a deep look using a Hex editor...

I will continue my quest...

**Comment by dmiller423** [comment]
Unfortunately, this type of behavior could be several different things ... and you have to take them all into account.   Do both images use the same sector size / raw?   Do both work in another emulator or in MESS?   It's possible they are different and it will help you find the answer.   It's also possible that someone did a bad conversion when upgrading it to newer chd format.   

Dump the full details of both,  image type , sector size, toc with sector count, filesize etc etc

**Comment by baka0815** [comment]
I created the CHD files myself using chdman 0.199 (or 0.200?) based on the working (and TOSEC verified) GDI image.

I may take a look at this the day after tomorrow or on Saturday.

**Comment by baka0815** [comment]
````c
blockoffs -= chd->header.hunkbytes / 2;
````
Works for Rayman 2 and F355, so they have to have something in common that differs to Crazy Taxi.

**Comment by baka0815** [comment]
Ok, I found out what's missing *phew*

We were missing the "extra frames":
````c
int padded = (frames + CD_TRACK_PADDING - 1) / CD_TRACK_PADDING;
track->extraframes = padded * CD_TRACK_PADDING - frames;
````
https://github.com/mamedev/mame/blob/63f29019c4e069c331f85e2c661935a57de036a9/src/lib/util/cdrom.cpp#L1214


Now I "just" need to wrap that all up and create a new PR...

**Pull request closed without merging**